accepted/single-file/design.md

```diff
@@ -0,0 +1,314 @@
+- For host/runtime support, the options are:
+  - Implement plugins using existing infrastructure. For example: Take control of assembly/native binary loads via existing `AssemblyLoadContext` callbacks and events. Extract the files embedded within the single-file plugin using the `GetFileStream()` API and load them on demand.
+  - Have new API to load a single-file plugin, for example: `AssemblyLoadContext.LoadWithEmbeddedDependencies()`.
+
+#### VS Integration
+
+Developers should be able to use the feature easily from Visual Studio. The feature will start with text based support -- by explicitly setting the `PublishSingleFile` property in the project file. In future, we may provide a single-file publish-profile, or other UI triggers.
+
+## User Experience
+
+To summarize, here's the overall experience for creating a HelloWorld single-file app 
+
+*  Create a new HelloWorld app: `HelloWorld$ dotnet new console`
+
+#### Framework Dependent HelloWorld
+
+* Normal publish: `dotnet publish` 
+
+  * Publish directory contains the host `HelloWorld.exe` ,  the app `HelloWorld.dll`, configuration files `HelloWorld.deps.json`, `HelloWorld.runtimeconfig.json`, and the PDB-file `HelloWorld.pdb`.
+
+* Single-file publish: `dotnet publish /p:PublishSingleFile=true`
+
+  * Publish directory contains: `HelloWorld.exe` `HelloWorld.pdb`
+
+  * `HelloWorld.dll`, `HelloWorld.deps.json`, and `HelloWorld.runtimeconfig.json` are embedded within `HelloWorld.exe`.
+
+* Run: `HelloWorld.exe`
+
+  * The app runs completely from the single-file, without the need for intermediate extraction to file.
+
+#### Self-Contained HelloWorld
+
+- Normal publish: `dotnet publish -r win10-x64 --self-contained`
+
+  * Publish directory contains 221 files including the host, the app, configuration files, the PDB-file and the runtime.
+
+- Single-file publish: `dotnet publish -r win10-x64 --self-contained /p:PublishSingleFile=true`
+
+  - Publish directory contains: `HelloWorld.exe` `HelloWorld.pdb`
+  - The remaining 219 files are embedded within the host `HelloWorld.exe`.
+
+- Run: `HelloWorld.exe`
+
+  * The bundled app and configuration files are processed directly from the bundle.
+  * Remaining 216 files will be extracted to disk at startup. 
+  * If reuse of extracted files is enabled, subsequent runs of the app may skip the extraction step.
+
+
+Most applications are expected to work without any changes. However, apps with a strong expectation about absolute location of dependent files may need to be made aware of bundling and extraction aspects of single-file publishing. No difference is expected with respect to debugging and analysis of apps.
+
+## Related Work
+
+* [Mono/MkBundle](https://github.com/mono/mono/blob/master/mcs/tools/mkbundle) 
+* [Fody.Costura](https://github.com/Fody/Costura)
+* [Warp](https://github.com/dgiagio/warp)
```

Comment:
Also: [dotnet-warp](https://github.com/Hubert-Rybak/dotnet-warp) global tool

(mentioned in https://www.hanselman.com/blog/BrainstormingCreatingASmallSingleSelfcontainedExecutableOutOfANETCoreApplication.aspx)